File: openclaw/src/device-code-setup.ts
// Plain Device Code (RFC 8628) plugin setup helper — no RAR.
//
// SSOT: docs/04_design/auth/device_code_setup_v1.md §3-2
//
// installTopology() (RAR 拡張版) と並列に、 RAR なしの単純な device code 経由 credential
// 取得を TS SDK で提供する。 backend は既存 RFC 8628 endpoint をそのまま reuse。
//
// Public API:
//   import {
//     setupViaDeviceCode,
//     type DeviceCodeSetupResult,
//     type DeviceCodeSetupPending,
//     InstallError, InstallDeniedError, InstallTimeoutError, InstallAuthError,
//     InstallAbortedError, InstallConfigError,
//   } from "./device-code-setup";

import {
  InstallAbortedError,
  InstallAuthError,
  InstallConfigError,
  InstallDeniedError,
  InstallError,
  InstallTimeoutError,
  parseOAuthError,
} from "./oauth-errors";
import { pollDeviceToken, safeFetch } from "./oauth-polling";

// Re-export error hierarchy for caller convenience (single import surface).
export {
  InstallAbortedError,
  InstallAuthError,
  InstallConfigError,
  InstallDeniedError,
  InstallError,
  InstallTimeoutError,
};

// ---------------------------------------------------------------------------
// Device credentials persistence (inline、 host file ~/.agentrux/device_credentials.json)
//
// 注: 既存 credentials.ts は plugins/ 配下で gitignore 対象 (.gitignore の *credentials*
// pattern)。 そのため DeviceCredentials の save/load helper は本 module に inline する。
// SSOT: docs/04_design/auth/device_code_setup_v1.md §4-1
// ---------------------------------------------------------------------------

import * as fs from "node:fs";
import * as path from "node:path";

export interface DeviceCredentials {
  base_url: string;
  dcr_client_id: string;
  access_token: string;
  refresh_token: string | null;
  issued_at_unix: number;
  expires_in: number;
  scope: string[];
  id_token?: string;
}

function agentruxDir(): string {
  const home = process.env.HOME || process.env.USERPROFILE || ".";
  return path.join(home, ".agentrux");
}

function deviceCredentialsPath(): string {
  return path.join(agentruxDir(), "device_credentials.json");
}

export function saveDeviceCredentials(creds: DeviceCredentials): void {
  const dir = agentruxDir();
  if (!fs.existsSync(dir)) {
    fs.mkdirSync(dir, { recursive: true, mode: 0o700 });
  }
  fs.writeFileSync(deviceCredentialsPath(), JSON.stringify(creds, null, 2), {
    mode: 0o600,
  });
}

export function loadDeviceCredentials(): DeviceCredentials | null {
  const filePath = deviceCredentialsPath();
  try {
    if (!fs.existsSync(filePath)) return null;
    const raw = JSON.parse(fs.readFileSync(filePath, "utf-8"));
    if (!raw || typeof raw !== "object") return null;
    if (
      typeof raw.dcr_client_id !== "string" ||
      typeof raw.access_token !== "string"
    ) {
      return null;
    }
    return raw as DeviceCredentials;
  } catch {
    return null;
  }
}

// Scope vocabulary (server is_valid_authorize_scope() と整合)
const VALID_SCOPE_VOCAB: ReadonlySet<string> = new Set([
  "topic.read",
  "topic.write",
  "openid",
  "email",
  "profile",
]);

const DEVICE_AUTHORIZE_PATH = "/oauth/device/authorize";
const USER_AGENT = "agentrux-sdk-device-code-setup-ts/1.0";
const DEFAULT_TIMEOUT_S = 600;
const MIN_TIMEOUT_S = 60;
const MAX_TIMEOUT_S = 600;

export interface DeviceCodeSetupPending {
  userCode: string;
  verificationUri: string;
  verificationUriComplete: string;
  expiresIn: number;
  interval: number;
}

export interface DeviceCodeSetupResult {
  accessToken: string;
  refreshToken: string | null;
  scope: string[];
  expiresIn: number;
  idToken?: string;
  grantedScopes: string[]; // alias of `scope` (legacy compat)
  grantedAtMs: number;
}

export type OnUserCode = (info: DeviceCodeSetupPending) => void | Promise<void>;

export interface SetupViaDeviceCodeOptions {
  baseUrl: string;
  clientId: string;
  scope?: readonly string[]; // default ["topic.read", "topic.write"]
  onUserCode?: OnUserCode;
  timeoutSeconds?: number; // default 600, clamped to [60, 600]
  signal?: AbortSignal;
  fetchImpl?: typeof fetch;
}

function rejectControlChars(value: string, field: string): void {
  for (let i = 0; i < value.length; i++) {
    const c = value.charCodeAt(i);
    if (c < 0x20 || c === 0x7f) {
      throw new InstallConfigError(`${field} contains control character`);
    }
  }
}

function validateScope(scope: readonly string[]): void {
  if (scope.length === 0) {
    throw new InstallConfigError("scope must be non-empty");
  }
  const seen = new Set<string>();
  for (const s of scope) {
    if (typeof s !== "string" || s.length === 0) {
      throw new InstallConfigError(
        `scope entry must be non-empty string: ${JSON.stringify(s)}`,
      );
    }
    if (seen.has(s)) {
      throw new InstallConfigError(`scope duplicate: ${JSON.stringify(s)}`);
    }
    seen.add(s);
    if (!VALID_SCOPE_VOCAB.has(s)) {
      throw new InstallConfigError(
        `scope ${JSON.stringify(s)} not in vocabulary ` +
          `(${JSON.stringify([...VALID_SCOPE_VOCAB].sort())})`,
      );
    }
    rejectControlChars(s, `scope[${JSON.stringify(s)}]`);
  }
}

export async function setupViaDeviceCode(
  opts: SetupViaDeviceCodeOptions,
): Promise<DeviceCodeSetupResult> {
  // 1. validate input
  if (
    typeof opts.baseUrl !== "string" ||
    !/^https?:\/\//.test(opts.baseUrl)
  ) {
    throw new InstallConfigError(`baseUrl must be http(s): ${opts.baseUrl}`);
  }
  if (typeof opts.clientId !== "string" || !opts.clientId) {
    throw new InstallConfigError("clientId is required");
  }
  rejectControlChars(opts.clientId, "clientId");
  const scope = opts.scope ?? ["topic.read", "topic.write"];
  validateScope(scope);

  const base = opts.baseUrl.replace(/\/+$/, "");
  const fetchImpl: typeof fetch =
    opts.fetchImpl ??
    (globalThis.fetch as typeof fetch | undefined) ??
    (() => {
      throw new InstallConfigError(
        "fetch is not available in this runtime (provide opts.fetchImpl)",
      );
    });
  const timeoutS = Math.max(
    MIN_TIMEOUT_S,
    Math.min(opts.timeoutSeconds ?? DEFAULT_TIMEOUT_S, MAX_TIMEOUT_S),
  );

  // 2. issue device_code
  const { deviceCode, pending } = await issueDeviceCode({
    base,
    clientId: opts.clientId,
    scope,
    fetchImpl,
    signal: opts.signal,
  });

  // 3. notify operator
  if (opts.onUserCode) {
    const cbRet = opts.onUserCode(pending);
    if (cbRet && typeof (cbRet as Promise<void>).then === "function") {
      await cbRet;
    }
  }

  // 4. poll /oauth/token (RFC 8628 §3.4、 oauth-polling 経由)
  const body = await pollDeviceToken({
    base,
    clientId: opts.clientId,
    deviceCode,
    userCode: pending.userCode,
    timeoutS,
    initialIntervalSeconds: pending.interval,
    fetchImpl,
    signal: opts.signal,
  });
  return parseTokenResponse(body);
}

interface IssueArgs {
  base: string;
  clientId: string;
  scope: readonly string[];
  fetchImpl: typeof fetch;
  signal?: AbortSignal;
}

async function issueDeviceCode(
  args: IssueArgs,
): Promise<{ deviceCode: string; pending: DeviceCodeSetupPending }> {
  const form = new URLSearchParams();
  form.set("client_id", args.clientId);
  form.set("scope", args.scope.join(" "));

  const res = await safeFetch(
    args.fetchImpl,
    `${args.base}${DEVICE_AUTHORIZE_PATH}`,
    {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": USER_AGENT,
        Accept: "application/json",
      },
      body: form.toString(),
      signal: args.signal,
    },
  );

  if (res.status === 200) {
    const body = (await res.json()) as Record<string, unknown>;
    return {
      deviceCode: String(body.device_code),
      pending: {
        userCode: String(body.user_code),
        verificationUri: String(body.verification_uri),
        verificationUriComplete: String(body.verification_uri_complete),
        expiresIn: Number(body.expires_in),
        interval: Number(body.interval),
      },
    };
  }

  const { code, desc } = await parseOAuthError(res);
  if (res.status === 400 && code === "invalid_client") {
    throw new InstallAuthError(`invalid_client: ${desc}`);
  }
  if (res.status === 400 && code === "invalid_scope") {
    throw new InstallAuthError(`invalid_scope: ${desc}`);
  }
  if (res.status === 400) {
    throw new InstallError(
      `${code || "invalid_request"}: ${desc}`,
      code || "invalid_request",
      400,
    );
  }
  if (res.status === 429) {
    throw new InstallError(`rate_limited: ${desc}`, "rate_limited", 429);
  }
  throw new InstallError(
    `unexpected status ${res.status}: ${desc}`,
    code,
    res.status,
  );
}

function parseTokenResponse(body: Record<string, unknown>): DeviceCodeSetupResult {
  if (typeof body.access_token !== "string" || !body.access_token) {
    throw new InstallError("malformed token response: access_token missing");
  }
  const expiresIn = Number(body.expires_in);
  if (!Number.isFinite(expiresIn)) {
    throw new InstallError("malformed token response: expires_in invalid");
  }
  const refreshTokenRaw = body.refresh_token;
  const refreshToken =
    typeof refreshTokenRaw === "string" && refreshTokenRaw
      ? refreshTokenRaw
      : null;
  const scopeRaw = body.scope;
  const scope =
    typeof scopeRaw === "string"
      ? scopeRaw.split(/\s+/).filter(Boolean)
      : [];
  const idTokenRaw = body.id_token;
  const idToken =
    typeof idTokenRaw === "string" && idTokenRaw ? idTokenRaw : undefined;

  return {
    accessToken: body.access_token,
    refreshToken,
    scope,
    expiresIn,
    idToken,
    grantedScopes: scope,
    grantedAtMs: Date.now(),
  };
}

File: sdk/src/agentrux/sdk/device_code_setup.py
"""Plain Device Code (RFC 8628) plugin setup helper — no RAR.

SSOT: docs/04_design/auth/device_code_setup_v1.md §3-1

`install_topology()` (RAR 拡張版) と並列に、 **RAR なし** の単純な device code 経由
credential 取得を SDK で提供する。 backend は既存 RFC 8628 endpoint (`POST /oauth/device/authorize`
+ `POST /device/verify` + `POST /oauth/token grant_type=device_code`) をそのまま reuse。

Public API:
    from agentrux.sdk.device_code_setup import (
        setup_via_device_code,
        DeviceCodeSetupResult,
        DeviceCodeSetupPending,
    )

Typical usage:
    result = await setup_via_device_code(
        base_url="https://api.agentrux.com",
        client_id="<dcr_client_id>",
        scope=("topic.read", "topic.write"),
        on_user_code=lambda info: print(f"Visit {info.verification_uri_complete}"),
    )
    print(result.access_token[:20], result.scope)

設計判断 (device_code_setup_v1.md §0-4):
- error 階層は既存 `InstallError` 系を **再利用** (新規 SetupError 階層不採用)
- scope vocab pre-validation (`topic.read|topic.write|openid|email|profile` のみ)
- `id_token` は openid scope 時のみ optional に含む
- polling は `_oauth_polling.poll_device_token()` を共有
"""

from __future__ import annotations

import inspect
import time
from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass, field
from typing import Any

import httpx

# Re-export Install* hierarchy (caller convenience、 device_code_setup_v1.md §3-1)
from agentrux.sdk._oauth_errors import (
    InstallAbortedError,
    InstallAuthError,
    InstallDeniedError,
    InstallError,
    InstallTimeoutError,
    parse_oauth_error,
)
from agentrux.sdk._oauth_errors import (
    InstallAuthError as _InstallAuthError,  # noqa: F401 re-export marker
)
from agentrux.sdk._oauth_errors import (
    InstallError as _InstallError,  # noqa: F401 re-export marker
)
from agentrux.sdk._oauth_polling import poll_device_token
from agentrux.sdk.errors import AgenTruxError, ConfigError

# Scope vocabulary (server `is_valid_authorize_scope()` と整合、 device_code_setup_v1.md §0-4)
_VALID_SCOPE_VOCAB: frozenset[str] = frozenset(
    {"topic.read", "topic.write", "openid", "email", "profile"}
)

_DEVICE_AUTHORIZE_PATH = "/oauth/device/authorize"
_USER_AGENT = "agentrux-sdk-device-code-setup/1.0"
_DEFAULT_TIMEOUT_S = 600
_MIN_TIMEOUT_S = 60


@dataclass(frozen=True, slots=True)
class DeviceCodeSetupPending:
    """`on_user_code` callback に渡る public 情報 (raw device_code は流さない)."""

    user_code: str
    verification_uri: str
    verification_uri_complete: str
    expires_in: int
    interval: int


@dataclass(frozen=True, slots=True)
class DeviceCodeSetupResult:
    """`setup_via_device_code()` の戻り値. Topology Flow `InstallResult` と shape は近いが、
    `authorization_details` 由来の topic_id_map / grants は含まない (plain device code は
    setup 時点で resource を作らないため)."""

    access_token: str                    # "aat_..."
    refresh_token: str | None            # "art_..." (public client なら必ず付く)
    scope: tuple[str, ...]               # granted scope vocabulary
    expires_in: int                      # access_token の TTL 秒
    id_token: str | None = None          # openid scope 指定時のみ
    granted_scopes: tuple[str, ...] = field(default_factory=tuple)  # scope alias (legacy compat)
    granted_at_unix: float = 0.0         # 受領時刻 (refresh_lead 計算用)


# callback can be sync or async
OnUserCodeCallback = Callable[[DeviceCodeSetupPending], None | Awaitable[None]]


def _validate_scope(scope: Sequence[str]) -> None:
    """Scope vocab 違反 / 重複 / 空 を ConfigError で reject (server に投げる前に early fail)."""
    if not scope:
        raise ConfigError("scope must be non-empty")
    seen: set[str] = set()
    for s in scope:
        if not isinstance(s, str) or not s:
            raise ConfigError(f"scope entry must be non-empty string: {s!r}")
        if s in seen:
            raise ConfigError(f"scope duplicate: {s!r}")
        seen.add(s)
        if s not in _VALID_SCOPE_VOCAB:
            raise ConfigError(
                f"scope {s!r} not in vocabulary "
                f"({sorted(_VALID_SCOPE_VOCAB)})"
            )
        # control char reject (server `_strip_validate_string` と整合)
        for ch in s:
            c = ord(ch)
            if c < 0x20 or c == 0x7F:
                raise ConfigError(f"scope {s!r} contains control character")


async def setup_via_device_code(
    *,
    base_url: str,
    client_id: str,
    scope: Sequence[str] = ("topic.read", "topic.write"),
    on_user_code: OnUserCodeCallback | None = None,
    timeout: int = _DEFAULT_TIMEOUT_S,  # noqa: ASYNC109 (caller-friendly seconds, not asyncio.timeout)
    connect_timeout: float = 5.0,
    read_timeout: float = 30.0,
    http: httpx.AsyncClient | None = None,
) -> DeviceCodeSetupResult:
    """RFC 8628 Device Authorization Grant (RAR なし) で credential を取得.

    Args:
        base_url: AgenTrux API base URL (例: "https://api.agentrux.com")。 trailing slash 許容。
        client_id: prior DCR で取得した OAuth public client UUID。
        scope: scope vocabulary (topic.read|topic.write|openid|email|profile の subset)。
            default は ("topic.read", "topic.write")。
        on_user_code: callback。 引数は DeviceCodeSetupPending。 同期 / async どちらも可。
        timeout: 全体 deadline 秒 (range [60, 600]、 device_code TTL ≤600s 制約)。
        connect_timeout / read_timeout: httpx timeouts。
        http: optional 共有 httpx.AsyncClient (caller 側で reuse する場合)。

    Returns:
        DeviceCodeSetupResult (aat_ + art_ + optional id_token)。

    Raises:
        ConfigError: input 不正 (client side detected: base_url / client_id / scope)。
        InstallAuthError: client_id 不正 / scope vocab 違反 / invalid_grant 系。
        InstallDeniedError: user が picker で deny。
        InstallTimeoutError: timeout / device_code expired。
        InstallAbortedError: asyncio.CancelledError catch (caller 側で raise する想定)。
        InstallError: その他致命的エラー。
        AgenTruxError: network エラー / 想定外 status。
    """
    if not isinstance(base_url, str) or not base_url.startswith(("http://", "https://")):
        raise ConfigError(f"base_url must be http(s) URL: {base_url!r}")
    if not isinstance(client_id, str) or not client_id:
        raise ConfigError("client_id is required")
    # control char reject (XSS like javascript: は startswith check で reject 済、
    # ここでは client_id の URL-unsafe char を guard)
    for ch in client_id:
        c = ord(ch)
        if c < 0x20 or c == 0x7F:
            raise ConfigError("client_id contains control character")
    _validate_scope(scope)

    base = base_url.rstrip("/")
    timeout_clamped = max(_MIN_TIMEOUT_S, min(int(timeout), _DEFAULT_TIMEOUT_S))

    if http is not None:
        # caller-provided client: do NOT close it
        return await _setup_with_client(
            http=http,
            base=base,
            client_id=client_id,
            scope=tuple(scope),
            on_user_code=on_user_code,
            timeout_s=timeout_clamped,
        )

    async with httpx.AsyncClient(
        base_url=base,
        timeout=httpx.Timeout(
            connect=connect_timeout, read=read_timeout, write=read_timeout, pool=read_timeout
        ),
        headers={"User-Agent": _USER_AGENT},
    ) as client:
        return await _setup_with_client(
            http=client,
            base=base,
            client_id=client_id,
            scope=tuple(scope),
            on_user_code=on_user_code,
            timeout_s=timeout_clamped,
        )


async def _setup_with_client(
    *,
    http: httpx.AsyncClient,
    base: str,
    client_id: str,
    scope: tuple[str, ...],
    on_user_code: OnUserCodeCallback | None,
    timeout_s: int,
) -> DeviceCodeSetupResult:
    """共有 httpx client を使った内部実装。 base_url 設定済前提."""
    # 1. issue device_code (POST /oauth/device/authorize)
    device_code, pending = await _issue_device_code(
        http=http, client_id=client_id, scope=scope
    )

    # 2. notify operator (sync / async callable 両対応)
    if on_user_code is not None:
        cb_ret = on_user_code(pending)
        if inspect.isawaitable(cb_ret):
            await cb_ret  # type: ignore[arg-type]

    # 3. poll /oauth/token (RFC 8628 §3.4、 _oauth_polling 経由)
    body = await poll_device_token(
        http=http,
        client_id=client_id,
        device_code=device_code,
        user_code=pending.user_code,
        timeout_s=timeout_s,
        initial_interval=float(pending.interval),
    )
    return _parse_token_response(body)


async def _issue_device_code(
    *,
    http: httpx.AsyncClient,
    client_id: str,
    scope: tuple[str, ...],
) -> tuple[str, DeviceCodeSetupPending]:
    """POST /oauth/device/authorize → (device_code, public pending info).

    device_code は polling 内部で使う bearer、 on_user_code には流さない (raw bearer の
    意図せぬログ流出を回避、 spec §3 設計判断)。
    """
    data = {
        "client_id": client_id,
        "scope": " ".join(scope),
    }
    try:
        r = await http.post(
            _DEVICE_AUTHORIZE_PATH,
            data=data,
            headers={"Content-Type": "application/x-www-form-urlencoded"},
        )
    except httpx.HTTPError as exc:
        raise AgenTruxError(f"network error during device authorize: {exc}") from exc

    if r.status_code == 200:
        body = r.json()
        return str(body["device_code"]), DeviceCodeSetupPending(
            user_code=str(body["user_code"]),
            verification_uri=str(body["verification_uri"]),
            verification_uri_complete=str(body["verification_uri_complete"]),
            expires_in=int(body["expires_in"]),
            interval=int(body["interval"]),
        )

    code, desc = parse_oauth_error(r)
    if r.status_code == 400 and code == "invalid_client":
        raise InstallAuthError(f"invalid_client: {desc}")
    if r.status_code == 400 and code == "invalid_scope":
        raise InstallAuthError(f"invalid_scope: {desc}")
    if r.status_code == 400:
        raise InstallError(f"{code or 'invalid_request'}: {desc}")
    if r.status_code == 429:
        raise InstallError(f"rate_limited: {desc}")
    raise AgenTruxError(
        f"unexpected status {r.status_code} from {_DEVICE_AUTHORIZE_PATH}: {desc}"
    )


def _parse_token_response(body: dict[str, Any]) -> DeviceCodeSetupResult:
    """200 OK token response → DeviceCodeSetupResult (plain device code 用、 RAR なし)."""
    try:
        access_token = str(body["access_token"])
        expires_in = int(body["expires_in"])
    except (KeyError, ValueError) as exc:
        raise InstallError(f"malformed token response: missing field {exc}") from exc

    refresh_token_raw = body.get("refresh_token")
    refresh_token = str(refresh_token_raw) if refresh_token_raw else None

    scope_raw = str(body.get("scope", ""))
    scope_tuple = tuple(scope_raw.split()) if scope_raw else ()

    id_token_raw = body.get("id_token")
    id_token = str(id_token_raw) if id_token_raw else None

    return DeviceCodeSetupResult(
        access_token=access_token,
        refresh_token=refresh_token,
        scope=scope_tuple,
        expires_in=expires_in,
        id_token=id_token,
        granted_scopes=scope_tuple,  # legacy alias
        granted_at_unix=time.time(),
    )


__all__ = [
    "DeviceCodeSetupPending",
    "DeviceCodeSetupResult",
    "InstallAbortedError",
    "InstallAuthError",
    "InstallDeniedError",
    "InstallError",
    "InstallTimeoutError",
    "OnUserCodeCallback",
    "setup_via_device_code",
]
